src/editors/playgroundResultProvider.ts

```diff
@@ -1,44 +1,141 @@
+  _playgroundResult: OutputItem;
+  _uri: vscode.Uri;
+
+  constructor(context: vscode.ExtensionContext) {
+    this._editDocumentCodeLensProvider = new EditDocumentCodeLensProvider();
+    this._playgroundResult = {
+      namespace: null,
+      type: null,
+      content: undefined
+    };
+    this._uri = vscode.Uri.parse('');
+
+    context.subscriptions.push(
+      vscode.languages.registerCodeLensProvider(
+        {
+          scheme: PLAYGROUND_RESULT_SCHEME,
+          language: 'json'
+        },
+        this._editDocumentCodeLensProvider
+      )
+    );
   }
 
   onDidChangeEmitter = new vscode.EventEmitter<vscode.Uri>();
   onDidChange = this.onDidChangeEmitter.event;
 
-  provideTextDocumentContent(): Promise<string> {
-    return new Promise((resolve) => {
-      this._statusView.showMessage('Getting results...');
-
-      if (
-        typeof this._playgroundController.playgroundResult?.content ===
-          'object' ||
-        this._playgroundController.playgroundResult?.type !== 'string'
-      ) {
-        return resolve(
-          JSON.stringify(
-            this._playgroundController.playgroundResult?.content,
-            null,
-            2
-          )
-        );
+  setPlaygroundResultUri(uri: vscode.Uri) {
+    this._uri = uri;
+  }
+
+  // When the playground evaluation response is received,
+  // use playgroundResult (namespace, type, content)
+  // to provide a virtual document content.
+  refreshPlaygroundResult(playgroundResult?: OutputItem): void {
+    if (playgroundResult) {
+      this._playgroundResult = playgroundResult;
+    }
+  }
+
+  // When the document is saved (Cmd/Ctrl + S or with the menu),
+  // update the document (content) in the collection
+  // or the single document to reflect the changes that were made.
+  refreshPlaygroundResultContent(document?: any): void {
+    if (
+      this._playgroundResult.content &&
+      Array.isArray(this._playgroundResult.content)
+    ) {
+      const content = [...this._playgroundResult.content];
+      const index = content.findIndex((item) => item._id === document._id);
+
+      if (typeof content[index] === 'object' && content[index] !== null) {
+        Object.keys(content[index]).forEach((item) => {
+          if (this._playgroundResult && this._playgroundResult.content) {
+            this._playgroundResult.content[index][item] = document[item];
+          }
+        });
       }
+    } else if (
+      this._playgroundResult.content !== null &&
+      typeof this._playgroundResult.content === 'object'
+    ) {
+      const content = { ...this._playgroundResult.content };
+
+      Object.keys(content).forEach((item) => {
+        if (this._playgroundResult && this._playgroundResult.content) {
+          this._playgroundResult.content[item] = document[item];
+        }
+      });
+    }
+
+    this.onDidChangeEmitter.fire(this._uri);
+  }
 
-      return resolve(this._playgroundController.playgroundResult?.content);
```

Comment:
This part potentially could cause https://jira.mongodb.org/browse/VSCODE-204 error, because provideTextDocumentContent expects Promise<string> result, and this._playgroundController.playgroundResult?.content could be undefined at some point. When manually tested different inputs and outputs of playground execution received once the same error that in the ticket. Refactored this function to check different result types and provided tests for them.